### ISSUE TITLE
[bug]:[v0.26.0rc][qwen3.5-122b-a10b-w8a8] Generation repetition loop on GSM8K

### ISSUE THREAD

**Issue by linnea-lin-00638949** [vllm-project/vllm-ascend#15166]
### Your current environment

| Item | Value |
| :--- | :--- |
| PyTorch | 2.10.0+cpu |
| torch_npu | 2.10.0.post4 |
| Python | 3.11.15 |
| vllm | 0.1.dev1+gd02df748b (commit `d02df748bf9efd99022f1a062597dc3cb3808485`, aligned via `vllm-ascend/.github/vllm-main-verified.commit`) |
| vllm-ascend | 0.1.dev1+g85625bd77 (branch `releases/v0.26.0rc`, commit `85625bd772f27a4a13bc8e548f131af5ce3ff734`) |
| OS | openEuler 24.03 (LTS) |
| CANN | Ascend-cann-toolkit 9.1.0 (innerversion V100R001C11SPC001B243) |
| Hardware | Ascend 910 (hpu910a3, 64GB HBM) × 8 (single node, TP=8) |

## Describe the Bug

Qwen3.5-122B-A10B-W8A8-MTP (a hybrid attention + recurrent/Mamba model, served with `--mamba-cache-mode align`) enters a **degenerate repetition loop during reasoning** on GSM8K questions and never emits a final answer.

The model reasons correctly to the answer (`460`), but then instead of stopping it repeats the phrase `I'll use \boxed{460}.` over and over until `max_tokens` is hit. The response returns with `finish_reason=length`, the `content` field is empty, and the entire `max_tokens` budget (32768 tokens) is consumed by the loop in the `reasoning` field.

Quantitatively, on a single request:
- `completion_tokens = 32768` (== max_tokens, looped to the cap)
- `finish_reason = length`
- The top 20-gram `ll use \boxed{460}` appears **1826 times** in the reasoning output
- The last ~400 chars of reasoning are pure repetition:

```
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
... (continues identically until max_tokens)
```

This is reproducible across consecutive requests (observed on multiple trials).

## To Reproduce

```bash
# --- environment ---
export HCCL_BUFFSIZE=1024
export VLLM_ASCEND_ENABLE_NZ=1
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export TASK_QUEUE_ENABLE=1
export HCCL_OP_EXPANSION_MODE=AIV
export OMP_NUM_THREADS=1

# --- launch ---
vllm serve <MODEL_PATH> \
  --trust-remote-code --served-model-name auto \
  --distributed-executor-backend mp \
  --model-loader-extra-config '{"enable_multithread_load":true,"num_threads":8}' \
  --enable-log-requests --enable-prompt-tokens-details \
  --port 8000 \
  -tp 8 -dp 1 \
  --max-model-len 262144 \
  --max-num-batched-tokens 16384 \
  --max-num-seqs 128 \
  --gpu-memory-utilization 0.90 \
  --compilation-config '{"cudagraph_capture_sizes":[4,16,32,64,96,128,160,192,224,256,288,320,352,384,416,448,480,512],"cudagraph_mode":"FULL_DECODE_ONLY"}' \
  --speculative-config '{"method":"qwen3_5_mtp","num_speculative_tokens":3,"enforce_eager":true}' \
  --async-scheduling \
  --quantization ascend \
  --enable-prefix-caching \
  --mamba-cache-mode align \
  --enable-expert-parallel \
  --reasoning-parser qwen3 \
  --enable-auto-tool-choice \
  --tool-call-parser qwen3_coder \
  --additional-config '{"enable_cpu_binding":true,"multistream_overlap_shared_expert":false}'

# --- trigger the repetition loop ---
curl -s http://localhost:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "auto",
    "messages": [{"role":"user","content":"Question: Eliza'\''s rate per hour for the first 40 hours she works each week is $10. She also receives an overtime pay of 1.2 times her regular hourly rate. If Eliza worked for 45 hours this week, how much are her earnings for this week?\nPlease reason step by step, and put your final answer within \\boxed{}."}],
    "max_tokens": 32768,
    "temperature": 0.6,
    "top_p": 0.95
  }'
# -> finish_reason=length, content=null, reasoning contains "I'\''ll use \\boxed{460}." repeated ~1800+ times
```

A non-repeating request (e.g. a short `hi`) returns normally with `finish_reason=stop`, so the server itself is healthy — the issue is specifically a generation/decoding repetition loop on longer reasoning prompts.

## Expected Behavior

The model should reason through the problem, emit the final answer inside `\boxed{}`, and **stop** (`finish_reason=stop`) well before `max_tokens`, producing a non-empty `content` with the answer. The repeated re-emission of the same preamble phrase until the token cap is a decoding anomaly, not expected model behavior.

## Additional Context

- **Model**: Qwen3.5-122B-A10B-W8A8-MTP — a hybrid model that uses recurrent/Mamba layers (served with `--mamba-cache-mode align`), so the `gated_delta_rule` recurrent kernel (`ChunkGatedDeltaRuleFwd` family, compiled into `_cann_ops_custom`) is on the execution path. The repetition loop is suspected to be related to this recurrent op's behavior; flagging here in case maintainers can correlate.
- **Speculative decoding**: Qwen3.5-MTP (`qwen3_5_mtp`, 3 draft tokens) is enabled. MTP mean acceptance length during the loop was ~2.95 (draft working), so the loop is in the main model output, not a draft mismatch.
- **Reproducibility**: Stable. Consecutive identical requests reproduce the loop with the same repeated phrase.
- **Impact / loop rate (100-sample GSM8K probe, same prompt + sampling, `max_tokens=4096`)**:
  - **42/100 (42%)** of generations hit `finish_reason=length` (looped to the token cap with the repeated phrase).
  - Of those 42 looping samples, only **15** had an extractable correct `\boxed{}` answer; **27** produced no usable final answer → wrong.
  - The 58 non-looping samples (finished normally) scored **57/58 ≈ 98%** correct.
  - Overall accuracy **72/100 = 72%** — i.e. the repeat loop costs roughly **~27 percentage points** of accuracy vs. the ~98% on non-looping samples, plus 42% of generations burn the full `max_tokens` budget. With the eval's `max_out_len=32768`, each looping sample takes ~5 min of generation, making a full eval run impractically slow (~hours).
- **max_tokens**: The loop only terminates because of `max_tokens`; without a cap the model would continue indefinitely.
- Not a crash: no traceback, no retCode — pure generation-quality regression.

**Comment by linnea-lin-00638949** [comment]
Adding the **raw verbatim model response** from the repro (same request as in "To Reproduce") so the exact loop pattern is visible, not paraphrased.

**Request** (the GSM8K question):
```
Question: Eliza's rate per hour for the first 40 hours she works each week is $10. She also receives an overtime pay of 1.2 times her regular hourly rate. If Eliza worked for 45 hours this week, how much are her earnings for this week?
Please reason step by step, and put your final answer within \boxed{}.
```
Settings: `max_tokens=32768, temperature=0.6, top_p=0.95`.

**Response metadata:**
- `finish_reason`: `length`
- `usage`: `{'prompt_tokens': 90, 'completion_tokens': 32768, 'total_tokens': 32858}` (completion == max_tokens cap)
- `message.content`: `null` (no final answer ever emitted)
- `message.reasoning`: 78546 chars long
- Top 20-gram `ll use \boxed{460}` appears **1826 times** in the reasoning

**Reasoning HEAD — first ~600 chars (the model reasons correctly to the answer):**
```
The user wants me to solve a word problem about calculating earnings.

**1. Identify the given information:**
*   **Regular hourly rate:** $10 per hour.
*   **Regular hours limit:** 40 hours per week.
*   **Overtime multiplier:** 1.2 times her regular hourly rate.
*   **Total hours worked:** 45 hours.

**2. Break down the calculation:**
*   **Step 1: Calculate regular earnings.**
    *   Eliza works 40 regular hours.
    *   Rate: $10/hour.
    *   Regular Earnings = 40 hours * $10/hour.

*   **Step 2: Calculate overtime hours.**
    *   Total hours = 45.
    *   Regular hours = 40.
    *   Ov
```

**Reasoning TAIL — last ~1500 chars (the degenerate loop, verbatim):**
```
boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

Okay, I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

I'll write the response now.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

Okay, I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

I'll write the response now.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

Okay, I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

I'll write the response now.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

Okay, I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.

I'll write the response now.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
I'll use `\boxed{460}`.
```

Notable: the model keeps emitting `I'll write the response now.` (declaring it's about to produce the final answer) and `Okay, I'll use \boxed{460}.`, but never actually transitions to writing the final `content` — it just re-emits this preamble block (~5 lines of `I'll use \boxed{460}`. separated by `I'll write the response now.` / `Okay, ...`) over and over until the token cap. The correct answer (`460`) is computed and present inside `\boxed{}` within the loop, but `content` stays `null`.

**Comment by linnea-lin-00638949** [comment]
Follow-up with an **ablation over the launch config** to check whether any flag combination avoids the repetition loop. (Trying to find a config that does not trigger the loop.)

## Method

For each variant: relaunch vllm-ascend with the modified config, then probe **15 GSM8K questions** (same prompt + sampling, `max_tokens=4096`, `temperature=0.6`, `top_p=0.95`; the first 15 of `gsm8k` `test.jsonl`, which includes the question from the original report). A response with `finish_reason=length` (hit the 4096 cap with the repeated phrase) is counted as a loop. Same baseline question set across all variants.

Baseline (the config in "To Reproduce") loops **5/15**.

## Results

| Variant | Disabled (vs baseline) | Loop / 15 | Note |
| :--- | :--- | :--- | :--- |
| baseline | — | 5 | original report config |
| V1 | `--speculative-config` (MTP) | 6 | MTP not the cause |
| V3 | MTP + `--async-scheduling` + `--enable-prefix-caching` + `--enable-expert-parallel` + `--additional-config` (cudagraph/quant/mamba/reasoning-parser kept on) | 4 | still loops |
| V4 | V3 + `--compilation-config`→`--enforce-eager` (no graph) + `--mamba-cache-mode` removed | 0 length, but 7/15 timed out at the probe's 300s | eager mode too slow — the looping samples just don't reach the 4096 cap within 300s; loop not actually gone |
| V5 | ENV only: `VLLM_ASCEND_ENABLE_NZ=0`, `TASK_QUEUE_ENABLE=0`, `HCCL_OP_EXPANSION_MODE` unset (all serve flags at baseline) | 4 | still loops |

The loop rate stays in the **27–42%** range (15-sample noise) across every variant — no meaningful change.

## Conclusion

The repetition loop **cannot be eliminated by any launch-config combination** tested. Flags ruled out as the cause: `--speculative-config` (MTP), `--compilation-config`/cudagraph (incl. `--enforce-eager`), `--async-scheduling`, `--enable-prefix-caching`, `--enable-expert-parallel`, `--additional-config`, `--mamba-cache-mode align`, and ENV `VLLM_ASCEND_ENABLE_NZ` / `TASK_QUEUE_ENABLE` / `HCCL_OP_EXPANSION_MODE`.

The only factors that remain on for every variant are:
- `--quantization ascend` (the checkpoint is W8A8; cannot be disabled without a different checkpoint), and
- the `gated_delta_rule` recurrent kernel (`ChunkGatedDeltaRuleFwd` family, in `_cann_ops_custom`) — the model is a hybrid attention + recurrent/Mamba model (`--mamba-cache-mode align`), so this kernel is always on the execution path.

This points at the recurrent kernel / quantization level rather than a launch flag. Consistent with the symptom (model reasons to the correct answer, then re-emits the `I'll use \boxed{...}.` preamble — including `I'll write the response now.` — without ever transitioning to final `content`), it looks like a decode-path anomaly in the recurrent state, not a sampling or graph issue.

**Comment by linnea-lin-00638949** [comment]
Follow-up with an isolation experiment that narrows the root cause further: **the repetition loop reproduces without any Ascend-specific code path — including on CPU with a pure-PyTorch reference implementation.**

## Setup

Same GSM8K-style prompt as the service-side repro (single problem, temp=0.6, top_p=0.95), offline generation, bf16 weights, ~1.5k–4k max_tokens.

## Isolation ladder

| # | Stack | Result |
|---|---|---|
| 1 | vLLM-Ascend + W8A8 (910B) | repeats (offline sampling, multiple seeds) |
| 2 | vLLM-Ascend + bf16 (910B) | repeats |
| 3 | HF `transformers` 5.2.0, pure-PyTorch fallback for the gated-delta-rule linear attention (`fla`/`causal_conv1d` not installed), bf16, on 910B | 1/3 sampled runs repeat |
| 4 | HF `transformers` 5.2.0, **CPU**, bf16 (`attn_implementation="eager"`, eager MoE experts — no Ascend dependency at all) | **2/2 sampled runs repeat** |

The failure signature is identical in all four: the model first computes the correct answer ($400 + $60 = $460), then gets stuck deliberating the final-answer boxing format, e.g.:

> "Final Answer construction: \boxed{460} or \boxed{\$460}. I'll stick to \boxed{460} or include the symbol. The prompt asks \"how much\", so \"$460\" is the most natural answer"

...until max_tokens. Greedy decoding mostly finishes normally; the loop is triggered by sampling, and its measured rate grows with max_tokens (more room for the loop to develop), which matches the original report's max_tokens dependence.

## Minimal CPU repro (no Ascend hardware needed)

```python
import torch
from transformers import AutoTokenizer, Qwen3_5MoeForConditionalGeneration

MP = "<bf16 checkpoint>"  # e.g. Qwen3.5-122B-A10B bf16
tok = AutoTokenizer.from_pretrained(MP)
model = Qwen3_5MoeForConditionalGeneration.from_pretrained(
    MP, dtype=torch.bfloat16, device_map="cpu",
    attn_implementation="eager",
)
for c in (model.config, getattr(model.config, "text_config", None)):
    if c is not None:
        c._experts_implementation = "eager"   # grouped_mm kernel has CPU alignment issues

prompt = tok.apply_chat_template(
    [{"role": "user", "content": "Question: Eliza's rate per hour for the first 40 hours she works each week is $10. She also receives an overtime pay of 1.2 times her regular hourly rate. If Eliza worked for 45 hours this week, how much are her earnings for this week?\nPlease reason step by step, and put your final answer within \\boxed{}."}],
    add_generation_prompt=True, tokenize=False)
ids = tok(prompt, return_tensors="pt")
out = model.generate(**ids, do_sample=True, temperature=0.6, top_p=0.95,
                     max_new_tokens=2048, pad_token_id=tok.eos_token_id)
print(tok.decode(out[0][ids["input_ids"].shape[1]:]))
```

With `transformers` 5.2.0 and no `fla` installed, the linear-attention layers run the pure-torch chunk/recurrent fallback, so the whole model is stock PyTorch. 2/2 sampled runs hit the deliberation loop on CPU.

## What this means

- Not specific to vLLM-Ascend, the Ascend kernels, or W8A8 quantization — it reproduces with the reference PyTorch implementation on CPU.
- Points to the model's sampling distribution itself (a strong attractor around re-deliberating the boxed final answer at temp 0.6), amplified by large max_tokens.
- Probably reproducible on upstream vLLM on other hardware as well; mitigations to evaluate are model/sampling-side (repetition/presence penalty — note penalty=1.1 was **not** sufficient in our offline test — or limiting the thinking budget), rather than engine flags.

Happy to share the full generated transcripts if useful.

**Comment by linnea-lin-00638949** [comment]
**Update: cross-stack confirmation — this is a model-side issue, not specific to vLLM-Ascend or Ascend hardware.**

We completed the isolation matrix for the single-prompt repro (GSM8K "Eliza" problem, temp 0.6, top_p 0.95):

| # | Stack | Hardware | Result |
|---|---|---|---|
| 1 | vLLM-Ascend 0.18.0–0.26.0 (service-side + offline, all config ablations) | Ascend 910B / 910B3 | repeats, 40–70% over 10 sampled runs |
| 2 | HF transformers 5.2.0, pure-torch gated-delta-rule fallback (no fla) | Ascend 910B | repeats (1/3) |
| 3 | HF transformers 5.2.0, pure-torch, eager attention + eager MoE | **CPU only** | repeats (2/2) |
| 4 | sglang (public-cloud GPU deployment of the same model) | **NVIDIA GPU** | **repeats — verbatim `\nI'll use \boxed{460}.` cycle to max_tokens (24576), `finish_reason=length`** |

The failure signature is identical on all four stacks: the model first computes the correct answer (40×$10 + 5×($10×1.2) = $460), then re-deliberates the boxed-answer format (`I will format it as \boxed{460}. Or \boxed{\$460}. ...`), and can degenerate into a verbatim `\nI'll use \boxed{460}.` loop that consumes the entire token budget.

Additional deterministic (greedy) evidence: with sampling removed, the vLLM-Ascend and pure-torch-CPU greedy traces are **token-identical for the first 215 tokens**, then diverge — one finishes cleanly, the other locks into the deliberation loop. So the loop is a strong attractor adjacent to the main trajectory; whether a run falls in depends on tiny numerical perturbations (quantization / kernel reduction order / sampling noise). This explains why every stack, hardware type, and quantization repeats at similar rates, why the rate grows with max_tokens, and why `repetition_penalty=1.1` does not suppress it.

**Conclusion: model capability issue** (Qwen3.5-122B under temp-0.6 sampling with long max_tokens), not an engine or hardware bug. No engine-side fix is expected from vLLM-Ascend; mitigation belongs to model/sampling configuration (stronger repetition/presence penalty, thinking-budget limits, or prompt-level constraints on the final answer format). Keeping this open for maintainer triage — happy to share full transcripts or the CPU repro script (comment above).

**Comment by linnea-lin-00638949** [comment]
模型能力问题，关闭
